### PR TITLE
Write stats to disk and `dolt_statistics` table

### PR DESCRIPTION
**Pull request by max-hoffman**
New flatbuffer type for statistics, which contains an address reference to a prolly.Map with a hardcoded statistics schema. Statistics for a table are rewritten every ANALYZE TABLE call.

Because the statistics schema is client-dependent, backwards compatibility will be fairly brittle right now. We could use a version identifier to invalidate a set of statistics if the client code creating a sql engine has a different version than the statistics read from disk.

### CODE REVIEW

File: go/libraries/doltcore/ref/stats_ref.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// StatsRefName is a dummy name, and there cannot be more than one stash ref.
+const StatsRefName = "stats"
+
+type StatsRef struct {
+	stash string
+}
+
+var _ DoltRef = StatsRef{}
+
+// NewStatsRef creates a reference to a stashes list. There cannot be more than one statsRef.
```
Comment (nicktobey) [inline go/libraries/doltcore/ref/stats_ref.go:26]: At first I thought that "stash" was a typo of "stats" but it's consistent enough that now I'm not sure. Can you clarify why we're calling these "stashes"?

Comment (max-hoffman) [reply go/libraries/doltcore/ref/stats_ref.go:26]: sorry bad copy paste, will fix

File: go/libraries/doltcore/doltdb/doltdb.go [hunk 2]
```diff
@@ -136,8 +137,8 @@ func (ddb *DoltDB) AccessMode() chunks.ExclusiveAccessMode {
 }
 
 // CommitRoot executes a chunkStore commit, atomically swapping the root hash of the database manifest
-func (ddb *DoltDB) CommitRoot(ctx context.Context, last, current hash.Hash) (bool, error) {
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/doltdb/doltdb.go:139]: yeah, it being reverse compared to the other commit interfaces it depends on is kind of confusing imo

[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/doltdb/doltdb.go:139]: Unless I'm missing something the change is all internal to this function call though, and makes it consistent with the other interfaces

File: go/serial/stat.fbs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+include "prolly.fbs";
+
+namespace serial;
+
+// StoreRoot is the top-level chunk in the store,
```
Comment (nicktobey) [inline go/serial/stat.fbs:19]: Replace this docstring?

File: go/serial/stat.fbs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+// containing references to all named Refs.
+table Statistic {
```
Comment (nicktobey) [inline go/serial/stat.fbs:21]: Is the plan to add more fields to this over time? Otherwise I'm not sure why the root doesn't just references the table address directly instead of going through this.

Unless I'm misunderstanding the purpose of this message type.

Comment (max-hoffman) [reply go/serial/stat.fbs:21]: Aaron and I talked about how it might be better to have a dedicated message type for this, otherwise the helper functions that limit the acceptable root types would have to include the generic prolly node message type.

**Comment by max-hoffman** [comment]
I realized last night that I need to tweak the update to not erase stats for other tables. Will add a fix where I delete all rows from the map for the given qualifier before re-adding.

File: go/libraries/doltcore/sqle/stats/stats_provider.go [hunk 5]
```diff
@@ -164,32 +172,83 @@ type indexMeta struct {
...
+// table stats in the database
+type dbStats struct {
+	db     string
+	active map[hash.Hash]int
+	stats  map[sql.StatQualifier]*DoltStats
 }
 
 var _ sql.StatsProvider = (*Provider)(nil)
 
+// Init scans the statistics tables, populating the |stats| attribute.
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/stats/stats_provider.go:201]: Init -> Load

File: go/libraries/doltcore/sqle/stats/stats_provider.go [hunk 6]
```diff
@@ -164,32 +172,83 @@ type indexMeta struct {
...
+				return err
+			}
+			stats, err := loadStats(ctx, dbName, m)
+			if errors.Is(err, ErrIncompatibleVersion) {
+				ctx.Warn(0, err.Error())
+				return nil
+			} else if err != nil {
+				return err
+			}
+			p.dbStats[dbName] = stats
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/stats/stats_provider.go:238]: It's not safe to modify maps concurrently: https://go.dev/blog/maps#TOC_6.

The official suggestion here is to use a mutex, but I suspect it would also be safe to insert nil values for each db name before creating any of the goroutines, and then reassigning those values in the goroutines. That should guarentee that the map doesn't get resized while another routine is writing to it.

File: go/libraries/doltcore/sqle/stats/write.go [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+	"github.com/dolthub/dolt/go/libraries/doltcore/schema"
+	"github.com/dolthub/dolt/go/store/prolly"
+	"github.com/dolthub/dolt/go/store/prolly/tree"
+	stypes "github.com/dolthub/dolt/go/store/types"
+	"github.com/dolthub/dolt/go/store/val"
+)
+
+func newStatsTable(ctx *sql.Context, ns tree.NodeStore, vrw stypes.ValueReadWriter) (*doltdb.Table, error) {
+	//  sqlutil.ToDoltSchema(ctx, root, tableName, sch, headRoot, collation)
+	return doltdb.CreateEmptyTable(ctx, ns, vrw, schema.StatsTableDoltSchema)
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/stats/write.go:33]: Remove this comment?

File: go/libraries/doltcore/sqle/stats/write.go [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+}
+
+// TODO: write a new set of statistics to the stats_provider
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/stats/write.go:36]: Clean up this docstring?

**Review by nicktobey — COMMENTED** [review]
Looks good!

I think the only thing missing is a test that we can write stats out to disk and read them back in again.

File: go/libraries/doltcore/schema/statistic.go [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+		&sql.Column{Name: statsVersion, Type: types.Int64},
+	},
+	PkOrdinals: []int{0, 1},
+}
+
+var StatsTableDoltSchema = StatsTableDoltSchemaGen()
+
+func StatsTableDoltSchemaGen() Schema {
+	colColl := NewColCollection(
+		NewColumn(statsQualifier, qualifierTag, stypes.StringKind, true, NotNullConstraint{}),
```
[the remark this answers is not in the document]
Comment (max-hoffman) [reply go/libraries/doltcore/schema/statistic.go:89]: Not super opinionated, seems like what other databases do, flexible for describing a physical path to the datastore for a set of buckets. It's maybe permissive if we want something other than db/table/index as the data source.

[the remark this answers is not in the document]
Comment (nicktobey) [reply go/libraries/doltcore/schema/statistic.go:89]: Sounds good.

File: go/libraries/doltcore/sqle/stats/stats_provider.go (hunk 6 above)
Comment (nicktobey) [reply go/libraries/doltcore/sqle/stats/stats_provider.go:241]: I think I misread; if p.dbStats[dbName] already exists for each dbName this is probably safe.

**Comment by max-hoffman** [comment]
I did some heavy organizational changes to the dolt stats PR, added `dolt_statistics` table, split out qualifier pieces into their own columns, added bunch of tests. Going to swing back to the GMS PR first to review, once that one's merged I'll do the Dolt side

File: go/libraries/doltcore/sqle/dtables/statistics_table.go [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+	"github.com/dolthub/dolt/go/store/val"
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/planbuilder"
+	stats2 "github.com/dolthub/go-mysql-server/sql/stats"
+	"strconv"
+	"strings"
+	"time"
+)
+
+// StatisticsTable is a sql.Table implementation that implements a system table which shows the dolt commit log
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/dtables/statistics_table.go:35]: "shows the dolt commit log" -> Update this?

File: go/libraries/doltcore/sqle/enginetest/stats_queries.go [hunk 11]
```diff
@@ -299,20 +303,83 @@ var DoltHistogramTests = []queries.ScriptTest{
...
+		SetUpScript: []string{
+			"CREATE table xy (x bigint primary key, y int, z varchar(500), key(y,z));",
+			"insert into xy values (0,0,'a'), (1,0,'a'), (2,0,'a'), (3,0,'a'), (4,1,'a'), (5,2,'a')",
+			"analyze table xy",
+		},
+		Assertions: []queries.ScriptTestAssertion{
+			{
+				Query: "select `database`, `table`, `index`, commit_hash, columns, types from dolt_statistics",
+				Expected: []sql.Row{
+					{"mydb", "xy", "PRIMARY", hash.Parse("f6la1u3ku5pucfctgrca2afq9vlr4nrs"), []string{"x"}, []string{"bigint"}},
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/enginetest/stats_queries.go:318]: I feel like hard coding the hash into the test makes it brittle and unreadable. The test should construct the value we expect it to point to and take the hash of that value.

File: go/libraries/doltcore/sqle/statistics_provider.go [hunk 12]
```diff
@@ -0,0 +1 @@
+package sqle
```
Comment (nicktobey) [inline go/libraries/doltcore/sqle/statistics_provider.go:1]: Is this file necessary?

**Review by nicktobey — APPROVED** [review]
Moving the stats logic into a custom `sql.Table` was a good idea.

Tests look good.

Approved.

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>6ca6950</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>6ca6950</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>6ca6950</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>b356903</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>b356903</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>b356903</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997676</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3e4bc44</TD>
<TD>not ok</TD>
<TD>138</TD>
</TR>
<TR><TD>3e4bc44</TD>
<TD>ok</TD>
<TD>5937319</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3e4bc44</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997676</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>96ee545</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>96ee545</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>96ee545</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>61bedc9</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>61bedc9</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>61bedc9</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>67dd08e</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>67dd08e</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>67dd08e</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.997709</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>5b0936c</TD>
<TD>not ok</TD>
<TD>136</TD>
</TR>
<TR><TD>5b0936c</TD>
<TD>ok</TD>
<TD>5937321</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>5b0936c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.997709</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>d6e28ed</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>d6e28ed</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>d6e28ed</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>cd83ac6</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>cd83ac6</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>cd83ac6</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@max-hoffman __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.997743 to 99.998097</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>305796c</TD>
<TD>not ok</TD>
<TD>113</TD>
</TR>
<TR><TD>305796c</TD>
<TD>ok</TD>
<TD>5937344</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>305796c</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.998097</TD>
</TR>
</table>
